File: codetraverse/ast_diff/purescriptdiff.py
from tree_sitter import Node
from .Detailedchanges import DetailedChanges
from .basefilediff import BaseFileDiff

class PureScriptFileDiff(BaseFileDiff):
    """Analyzes and compares two PureScript ASTs for semantic differences."""

    def __init__(self, module_name=""):
        self.changes = DetailedChanges(module_name)

    def get_decl_name(self, node: Node) -> str:
        """Finds the name of a PureScript declaration."""
        # Case 1: Standard declarations with a 'name' field
        name_node = node.child_by_field_name('name')
        if name_node:
            return name_node.text.decode('utf8')

        # Case 2: Function declarations (name is the first identifier)
        if node.type == 'function':
            if node.child_count > 0 and node.children[0].type == 'identifier':
                return node.children[0].text.decode('utf8')

        # Case 3: Type aliases (e.g., `type String = ...`)
        if node.type == 'type_alias_declaration':
            # The structure is typically `type_alias_declaration -> type_variable_binding -> type_identifier`
            tvb_node = next((c for c in node.children if c.type == 'type_variable_binding'), None)
            if tvb_node:
                name_node = next((c for c in tvb_node.children if c.type == 'type_identifier'), None)
                if name_node:
                    return name_node.text.decode('utf8')

        # Case 4: Foreign imports (e.g., `foreign import foo :: Int`)
        if node.type == 'foreign_import':
            name_node = next((c for c in node.children if c.type == 'identifier'), None)
            if name_node:
                return name_node.text.decode('utf8')

        # Case 5: Class instances (e.g., `instance showInt :: Show Int`)
        if node.type == 'class_instance':
            # Create a composite name like "Show Int"
            instance_name_node = node.child_by_field_name('instance_name')
            if instance_name_node:
                # The instance name itself can be complex, so we take its full text
                return instance_name_node.text.decode('utf8').strip()

        return None

    def extract_components(self, root: Node):
        """Extracts all declarations from a PureScript AST using recursive traversal."""
        items = {
            "functions": {}, "classes": {}, "data_declarations": {},
            "newtypes": {}, "type_aliases": {}, "foreign_imports": {},
            "instances": {},
        }

        node_type_map = {
            "function": items["functions"],
            "class_declaration": items["classes"],
            "data_declaration": items["data_declarations"],
            "newtype": items["newtypes"],
            "type_alias_declaration": items["type_aliases"],
            "foreign_import": items["foreign_imports"],
            "class_instance": items["instances"],
        }

        queue = [root]
        while queue:
            current_node = queue.pop(0)
            if current_node.type in node_type_map:
                name = self.get_decl_name(current_node)
                if name:
                    target_dict = node_type_map[current_node.type]
                    target_dict[name] = (current_node, current_node.text.decode('utf8'), current_node.start_point, current_node.end_point)
            
            # Continue traversal
            for child in current_node.children:
                queue.append(child)

        return items

    def deep_equal(self, nodeA: Node, nodeB: Node) -> bool:
        """Recursively checks if two AST nodes are structurally identical."""
        if (nodeA is None) != (nodeB is None):
            return False
        if nodeA is None and nodeB is None:
            return True
        if nodeA.type != nodeB.type:
            return False
        
        # For leaf nodes, compare text content
        if not nodeA.children:
            return nodeA.text == nodeB.text

        if len(nodeA.children) != len(nodeB.children):
            return False

        for childA, childB in zip(nodeA.children, nodeB.children):
            if not self.deep_equal(childA, childB):
                return False

        return True

    def diff_components(self, before_map: dict, after_map: dict):
        """Compares two dictionaries of components using deep equality and returns the diff."""
        before_names = set(before_map.keys())
        after_names = set(after_map.keys())
        
        added_names = after_names - before_names
        deleted_names = before_names - after_names
        common_names = before_names & after_names
        
        added = [(n, after_map[n][1], {"start": after_map[n][2], "end": after_map[n][3]}) for n in sorted(added_names)]
        deleted = [(n, before_map[n][1], {"start": before_map[n][2], "end": before_map[n][3]}) for n in sorted(deleted_names)]
        
        modified = []
        for name in sorted(common_names):
            old_ast, old_body, old_start, old_end = before_map[name]
            new_ast, new_body, new_start, new_end = after_map[name]
            
            # Use deep_equal for robust comparison
            if not self.deep_equal(old_ast, new_ast):
                 modified.append((name, old_body, new_body, {"old_start": old_start, "old_end": old_end, "new_start": new_start, "new_end": new_end}))
        
        return {"added": added, "deleted": deleted, "modified": modified}

    def compare_two_files(self, old_file_ast: Node, new_file_ast: Node) -> DetailedChanges:
        """The main method to compare two PureScript files."""
        old_items = self.extract_components(old_file_ast.root_node)
        new_items = self.extract_components(new_file_ast.root_node)

        # Iterate over all component categories (functions, classes, types, etc.)
        all_categories = set(old_items.keys()) | set(new_items.keys())

        for category in all_categories:
            old_map = old_items.get(category, {})
            new_map = new_items.get(category, {})
            
            diff = self.diff_components(old_map, new_map)
            
            for change_type in ["added", "deleted", "modified"]:
                for data in diff[change_type]:
                    self.changes.add_change(category, change_type, data)

        return self.changes
    
    def process_single_file(self, file_ast: Node, mode="deleted") -> DetailedChanges:
        """Processes a single file that was either entirely added or deleted."""
        items = self.extract_components(file_ast.root_node)

        for category, component_map in items.items():
            for name, data_tuple in component_map.items():
                item = (name, data_tuple[1], {"start": data_tuple[2], "end": data_tuple[3]})
                self.changes.add_change(category, mode, item)
        
        return self.changes

File: codetraverse/ast_diff/purescriptdiff.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PureScriptFileDiff = void 0;
var Detailedchanges_1 = require("./Detailedchanges");
var basefilediff_1 = require("./basefilediff");
var PureScriptFileDiff = /** @class */ (function (_super) {
    __extends(PureScriptFileDiff, _super);
    function PureScriptFileDiff(moduleName) {
        if (moduleName === void 0) { moduleName = ""; }
        var _this = _super.call(this, moduleName) || this;
        _this.changes = new Detailedchanges_1.DetailedChanges(moduleName);
        return _this;
    }
    PureScriptFileDiff.prototype.getDeclName = function (node) {
        var nameNode = node.childForFieldName('name');
        if (nameNode) {
            return nameNode.text;
        }
        if (node.type === 'function') {
            if (node.childCount > 0 && node.children[0].type === 'identifier') {
                return node.children[0].text;
            }
        }
        if (node.type === 'type_alias_declaration') {
            var tvbNode = node.children.find(function (c) { return c.type === 'type_variable_binding'; });
            if (tvbNode) {
                var nameNode_1 = tvbNode.children.find(function (c) { return c.type === 'type_identifier'; });
                if (nameNode_1) {
                    return nameNode_1.text;
                }
            }
        }
        if (node.type === 'foreign_import') {
            var nameNode_2 = node.children.find(function (c) { return c.type === 'identifier'; });
            if (nameNode_2) {
                return nameNode_2.text;
            }
        }
        if (node.type === 'class_instance') {
            var instanceNameNode = node.childForFieldName('instance_name');
            if (instanceNameNode) {
                return instanceNameNode.text.trim();
            }
        }
        return null;
    };
    PureScriptFileDiff.prototype.extractComponents = function (root) {
        var items = {
            "functions": {}, "classes": {}, "data_declarations": {},
            "newtypes": {}, "type_aliases": {}, "foreign_imports": {},
            "instances": {},
        };
        var nodeTypeMap = {
            "function": items.functions,
            "class_declaration": items.classes,
            "data_declaration": items.data_declarations,
            "newtype": items.newtypes,
            "type_alias_declaration": items.type_aliases,
            "foreign_import": items.foreign_imports,
            "class_instance": items.instances,
        };
        var queue = [root];
        while (queue.length > 0) {
            var currentNode = queue.shift();
            if (currentNode.type in nodeTypeMap) {
                var name_1 = this.getDeclName(currentNode);
                if (name_1) {
                    var targetDict = nodeTypeMap[currentNode.type];
                    targetDict[name_1] = [currentNode, currentNode.text, currentNode.startPosition, currentNode.endPosition];
                }
            }
            for (var _i = 0, _a = currentNode.children; _i < _a.length; _i++) {
                var child = _a[_i];
                queue.push(child);
            }
        }
        return items;
    };
    PureScriptFileDiff.prototype.deepEqual = function (nodeA, nodeB) {
        if (!nodeA || !nodeB) {
            return nodeA === nodeB;
        }
        if (nodeA.type !== nodeB.type) {
            return false;
        }
        if (nodeA.children.length === 0) {
            return nodeA.text === nodeB.text;
        }
        if (nodeA.children.length !== nodeB.children.length) {
            return false;
        }
        for (var i = 0; i < nodeA.children.length; i++) {
            if (!this.deepEqual(nodeA.children[i], nodeB.children[i])) {
                return false;
            }
        }
        return true;
    };
    PureScriptFileDiff.prototype.diffComponents = function (beforeMap, afterMap) {
        var beforeNames = new Set(Object.keys(beforeMap));
        var afterNames = new Set(Object.keys(afterMap));
        var addedNames = Array.from(afterNames).filter(function (name) { return !beforeNames.has(name); });
        var deletedNames = Array.from(beforeNames).filter(function (name) { return !afterNames.has(name); });
        var commonNames = Array.from(beforeNames).filter(function (name) { return afterNames.has(name); });
        var added = addedNames.sort().map(function (n) { return [n, afterMap[n][1], { start: afterMap[n][2], end: afterMap[n][3] }]; });
        var deleted = deletedNames.sort().map(function (n) { return [n, beforeMap[n][1], { start: beforeMap[n][2], end: beforeMap[n][3] }]; });
        var modified = [];
        for (var _i = 0, _a = commonNames.sort(); _i < _a.length; _i++) {
            var name_2 = _a[_i];
            var _b = beforeMap[name_2], oldAst = _b[0], oldBody = _b[1], oldStart = _b[2], oldEnd = _b[3];
            var _c = afterMap[name_2], newAst = _c[0], newBody = _c[1], newStart = _c[2], newEnd = _c[3];
            if (!this.deepEqual(oldAst, newAst)) {
                modified.push([name_2, oldBody, newBody, { old_start: oldStart, old_end: oldEnd, new_start: newStart, new_end: newEnd }]);
            }
        }
        return { added: added, deleted: deleted, modified: modified };
    };
    PureScriptFileDiff.prototype.compareTwoFiles = function (oldFileAst, newFileAst) {
        var oldItems = this.extractComponents(oldFileAst);
        var newItems = this.extractComponents(newFileAst);
        var allCategories = new Set(__spreadArray(__spreadArray([], Object.keys(oldItems), true), Object.keys(newItems), true));
        for (var _i = 0, _a = Array.from(allCategories); _i < _a.length; _i++) {
            var category = _a[_i];
            var oldMap = oldItems[category] || {};
            var newMap = newItems[category] || {};
            var diff = this.diffComponents(oldMap, newMap);
            for (var changeType in diff) {
                for (var _b = 0, _c = diff[changeType]; _b < _c.length; _b++) {
                    var data = _c[_b];
                    this.changes.add_change(category, changeType, data);
                }
            }
        }
        return this.changes;
    };
    PureScriptFileDiff.prototype.processSingleFile = function (fileAst, mode) {
        if (mode === void 0) { mode = "deleted"; }
        var items = this.extractComponents(fileAst);
        for (var category in items) {
            var componentMap = items[category];
            for (var name_3 in componentMap) {
                var dataTuple = componentMap[name_3];
                var item = [name_3, dataTuple[1], { start: dataTuple[2], end: dataTuple[3] }];
                this.changes.add_change(category, mode, item);
            }
        }
        return this.changes;
    };
    return PureScriptFileDiff;
}(basefilediff_1.BaseFileDiff));
exports.PureScriptFileDiff = PureScriptFileDiff;
